git_utils_shared/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+
+    let merge_base = String::from_utf8(merge_base_vec).unwrap();
+    let merge_base = merge_base.trim().to_owned();
+
+    if merge_base.trim().is_empty() {
+        return None;
+    } else {
+        return Some(merge_base);
+    }
+}
+
+/// Show commits on the second branch that are not on the first branch
+pub fn show_uncommon_commit_from_other_branch(branch: &str, other_branch: &str) {
+    let _cmd1 = Command::new("git")
+        .arg("--no-pager")
+        .arg("log")
+        .arg("--pretty=format:%Cgreen%h%Creset %Cred(%an)%Creset [%ad] %Cblue%s%Creset ")
+        .arg(branch)
+        .arg(format!("^{}", other_branch))
+        .status()
+        .expect("Failed to retrieve unique commits");
+    println!("\n");
+}
+
+/// Show the common commit
+pub fn show_common_commit(merge_base: &str) {
+    let _cmd3 = Command::new("git")
+        .arg("--no-pager")
+        .arg("log")
+        .arg("-1")
+        .arg(&merge_base)
+        .status()
+        .expect("Failed to get first common commit");
+}
+
+pub fn get_files_with_word(search: &str, path: &str) -> Option<Vec<String>> {
+    let cmd = Command::new("git")
+        .args([
+            "--no-pager",
+            "grep",
+            "--full-name",
+            "--name-only",
+            search,
+            path,
+        ])
+        .output()
+        .expect("Failed")
+        .stdout;
+    let stdout = String::from_utf8(cmd).unwrap();
+    if stdout.is_empty() {
+        return None;
+    }
+
+    Some(
+        stdout
+            .trim()
+            .split("\n")
+            .map(|path| path.to_owned())
+            .collect::<Vec<String>>(),
+    )
+}
+
+pub fn get_files_with_word_using_grep(search: &str, path: &str) -> Option<Vec<String>> {
+    let cmd = Command::new("grep")
+        .args([search, path, "-lR"])
+        .output()
+        .expect("Failed")
+        .stdout;
+    let stdout = String::from_utf8(cmd).unwrap();
+    if stdout.is_empty() {
+        return None;
+    }
+    Some(
+        stdout
+            .trim()
+            .split("\n")
+            .map(|path| path.to_owned().replace("./", ""))
+            .collect::<Vec<String>>(),
+    )
+}
+
+pub mod file {
+    use std::process::Command;
+    pub fn file_is_modified(file_path: &str) -> bool {
+        !Command::new("git")
+            .args(["diff", "--quiet", &file_path])
+            .stdout(std::process::Stdio::null())
+            .stderr(std::process::Stdio::null())
+            .status()
+            .expect("Failed ")
+            .success()
+    }
+
+    pub fn file_is_tracked(file_path: &str) -> bool {
+        Command::new("git")
+            .args(["ls-files", "--error-unmatch", &file_path])
+            .stdout(std::process::Stdio::null())
+            .stderr(std::process::Stdio::null())
+            .status()
+            .expect("Failed ")
+            .success()
+    }
+
+    pub fn file_is_untracked(file_path: &str) -> bool {
+        !file_is_tracked(file_path)
+    }
+}
```

Comment:
Unfortunately, this requires an iteration on a per-file basis, which can be slow on large repositories.  

Instead, at the beginning of the command, run a `git status --porcelain=v2` and all at once put the new or modified files in a `Set` for further fast lookups.

If content has been added to the index (e.g. `git add`) then it should be considered clean and elegible for `git srep` doing a replacement on it.

That supports the following workflow:

1. `git srep foo bar .`
2. review `git diff`
3. `git add .`
4. `git srep bar baz .`  (no error here because the working tree is clean)
5. review `git diff`
6. `git checkout`, if you don't like it, etc...